cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java

```diff
@@ -1107,6 +1107,31 @@ private RunRecordMeta getCompletedRun(MDSKey completedKey, final String runid) {
     return newRecords;
   }
 
+  /**
+   * Fetches the historical (i.e COMPLETED or FAILED or KILLED) run records from a given set of namespaces.
+   *
+   * @param namespaces fetch run history that is belonged to one of these namespaces
+   * @param earliestStopTime fetch run history that has stopped at or after the earliestStopTime in seconds
+   * @param latestStartTime fetch run history that has started before the latestStartTime in seconds
+   * @param limit max number of entries to fetch for this history call
+   * @return map of logged runs
+   */
+  public Map<ProgramRunId, RunRecordMeta> getHistoricalRuns(final Set<String> namespaces,
+                                                            final long earliestStopTime, final long latestStartTime,
+                                                            final int limit) {
+    MDSKey keyPrefix = new MDSKey.Builder().add(TYPE_RUN_RECORD_COMPLETED).build();
+    //return all records in each namespace
+    Map<ProgramRunId, RunRecordMeta> runs = new HashMap<>();
+    namespaces.stream()
+      .map(ns -> getProgramRunIdMap(listKV(new MDSKey.Builder(keyPrefix).add(ns).build(), null,
+                                           RunRecordMeta.class, limit, key -> true,
+                                           // only get runs with stop time equal to or later than the earliestStopTime
+                                           // and start time earlier than latestStartTime
+                                           meta -> meta.getStopTs() != null && meta.getStopTs() >= earliestStopTime
+                                             && meta.getStartTs() < latestStartTime))).forEach(runs::putAll);
```

Comment:
i know most of this will be rewritten, but will be good to explain the logic in detail in comments, so will help in future to understand easily.  something like
``` we would like to get active runs in a time window, the beginning of this time window is the earliestStopTime and end of this time window is the latestStartTime, we want to exclude program run records that stopped before the begin of the window (earliestStopTime) and program run records that started after the end of the window (latestStartTime), all other records will be active at some point in the time window and will be included ```